### PR TITLE
IPPM graph refactor

### LINKED ISSUE #362: Refine IPPM graph datastructure
There is some duplication in the IPPM code, stemming from it's original design against the API. This means that some graphing code, and some classes which define hexels, spikes, pairings, etc, are similar, and could probably be consolidated. Also, I (@caiw) get a little confused when reading through how the graphs datastructure is currently built and traversed (it's not wrong or bad, I just wonder if it could be more intuitively defined).

Low priority, but I've been wondering about combining it into a `networkx.DiGraph`, perhaps whose nodes are literal `Transform` objects (see #65).

We can discuss pros and cons on this issue.

### PR DESCRIPTION
**Pull request by caiw**
Fixes #362

### CODE REVIEW

**Comment by caiw** [comment]
Feature complete! Linting passes! Now to make the rest of the tests pass...

**Comment by caiw** [comment]
And all the changes are improving things, adding new tests, improving comments. Testing rules! Thanks AGAIN to @anirudh1666 for setting up such a comprehensive suite for this code. 🙏

**Comment by caiw** [comment]
Note to self to speed up: Instead of clearing 10s of 1000s of points, simply filter the few you want to keep - it'll be much faster, and will use nearly the same numpy code!

**Comment by neukym** [comment]
Amazing. @anirudh1666 - can you take a first look?

**Comment by caiw** [comment]
Oof.. it's really slow. I did some optimisations this morning but it's still noticeably slower than the original code - something to work on with proper profiling.

Also I confirmed (by running the demo) that the failing tests are a function of the denoising not working as intended. I'm sure it's an easy fix as I was only changing data structures not algorithms - probably just something silly. Still sorry to be giving you broken code to read @anirudh1666 - but I'd still really value your input at this stage.

File: kymata/ippm/ippm.py [hunk 1]
```diff
@@ -1,19 +1,100 @@
...
+                raise ValueError(denoiser)
+        else:
+            denoising_strategy = None
+
+        # Do the denoising
+        if denoising_strategy is not None:
+            expression_set = denoising_strategy.denoise(expression_set)
+
+        # Build the graph
+        self._graphs: dict[str, IPPMGraph] = dict()
```
Comment (anirudh1666) [inline kymata/ippm/ippm.py:86]: nit: for clarity, it may help to place self._graphs near the definition in the comments. Usually the constructor contains all of the attributes for a class in one place for easy reference

File: kymata/entities/expression.py [hunk 2]
```diff
@@ -5,26 +5,24 @@
 from __future__ import annotations
```
Comment (anirudh1666) [inline kymata/entities/expression.py:5]: What is the purpose of densify_data_block in this file?

File: kymata/ippm/denoising_strategies.py [hunk 3]
```diff
@@ -1,375 +1,410 @@
...
         self,
-        hemi: str,
-        should_normalise: bool = False,
-        should_cluster_only_latency: bool = False,
-        should_max_pool: bool = False,
-        normal_dist_threshold: float = 5,
-        should_merge_hemis: bool = False,
-        should_exclude_insignificant: bool = True,
-        should_shuffle: bool = True,
+        should_normalise: bool,
```
Comment (anirudh1666) [inline kymata/ippm/denoising_strategies.py:28]: It might be worthwhile to add some default values for these parameters. It comes down to whether you think which parameters have high variance across runs and which ones will remain constant. The constant ones can be moved to optional/default valuse.

Comment (anirudh1666) [reply kymata/ippm/denoising_strategies.py:28]: I think should_normalise, should_cluster_only_latency, should_max_pool will be False for most runs unless someone wants to experiment.

should_shuffle will be True. As for the threshold for significance, not sure, although estimating it from the data seems better than absolute threshold (exclude_points_above_n_sigma > exclude_logp_vals_above)

File: kymata/ippm/ippm.py (hunk 1 above)
Comment (caiw) [reply kymata/ippm/ippm.py:86]: When initialising a variable and then conditionally mutating it, I like to keep all that in one place so it's clear when reading the code mutating value that the variable being mutated doesn't have a prior value set somewhere above.  However I've also added a comment which mentions the structure of the `self._graphs` dict right there. It's a repetition from the docstring, which isn't ideal, but I personally think keeps things clearer.

File: kymata/entities/expression.py (hunk 2 above)
Comment (caiw) [reply kymata/entities/expression.py:5]: When searching for best transforms (i.e. doing model selection) we need to do an argmin, but the n-dim sparse matrix object we're using (`sparse.COO`) doesn't implement `argmin` (or actually `idxmin` as `xarray` calls it), so I have to temporarily convert to a `numpy.ndarray`.  Not ideal, but I can't see an alternative beyond re-implementing argmin for `sparse.COO`.

File: kymata/ippm/denoising_strategies.py (hunk 3 above)
Comment (caiw) [reply kymata/ippm/denoising_strategies.py:28]: I've moved the existing default values to the subclasses, but it sounds like it makes logical sense for them to be set in the base class too. Having said that, the existing defaults weren't quite what you said here, so I will adjust the default values in the subclasses to what you've said here, unless they had a comment saying why they should be different.

Comment (caiw) [reply kymata/ippm/denoising_strategies.py:28]: For now I've left the defaults for the `should exclude_` params as `None` (i.e. no exclusion) but we can revisit that.

**Comment by neukym** [comment]
demo_ippm.ipynb isn't working quite right for me: This seems ... partially denoised?
Screenshot 2025-02-12 at 10 13 29

**Comment by caiw** [comment]
@neukym Yeah I saw this, it is related to the non-passing tests. Something trivial I think but I've not tracked down what yet.

**Comment by caiw** [comment]
Here's a record of where I got to before signing off this evening. @anirudh1666 perhaps this will help as you look at the tests; and I'll pick up again here next week.

I've been stepping through `demo_ippm.ipynb` on both `main` and this branch.  As @neukym [noted above](https://github.com/kymata-atlas/kymata-core/pull/391#issuecomment-2653269638), the denoising "isn't working" on this branch.

In particular, here's how the full "Kymata2023Q3" dataset looks like when denoised on main:
clipboard
and on branch:
clipboard11

My plan was to simplify this comparison by just filtering down to one transform at the top of the notebook using
```python
expression_set: HexelExpressionSet = KymataMirror2023Q3Dataset().to_expressionset()
expression_set = expression_set[[
    "TVL loudness (short-term)",
]]
```

Now here's something interesting.

On main, this looks as expected:
clipboard1
And now on branch:
clipboard2

**So when filtering down to one function, the branch works just like main.**

Now moving up to two functions.

```python
expression_set: HexelExpressionSet = KymataMirror2023Q3Dataset().to_expressionset()
expression_set = expression_set[[
    "TVL loudness (instantaneous)",
    "TVL loudness (short-term)",
]]
```

On main:
clipboard8
On branch:
clipboard9

So somehow the denoising is changing when adding multiple functions, on branch.

Next comment is one difference I've found, but it needs some discussion/thinking.

**Comment by caiw** [comment]
## How are points filtered for significance?

On main, points are filtered for significance using a threshold set in the strategy init:

```python
class DenoisingStrategy(ABC):
    """
    Superclass for unsupervised clustering algorithms.
    Strategies should conform to this interface.
    """

    def __init__(self, ..., should_exclude_insignificant: bool = True, ...):
        ...
        self._threshold_for_significance = (
            DenoisingStrategy._estimate_threshold_for_significance(
                normal_dist_threshold  # defaults to 5; i.e. 5-sigma thresholding
            )
        )
```

Here's the entirety of `estimate_threshold_for_significance()`, minus the docstring:

```python
    @staticmethod
    def _estimate_threshold_for_significance(x: float) -> float:
        def __calculate_bonferroni_corrected_alpha(one_minus_probability):
            return 1 - pow(
                (1 - one_minus_probability), (1 / (2 * TIMEPOINTS * NUMBER_OF_HEXELS))
            )

        probability_X_gt_x = 1 - NormalDist(mu=0, sigma=1).cdf(x)
        threshold_for_significance = __calculate_bonferroni_corrected_alpha(
            probability_X_gt_x
        )
        return threshold_for_significance
```

A quick note here that this is called "bonferroni" because the code was copied from an old version of the codebase before @ollieparish rightly pointed out this is in fact Šidák correction not Bonferroni correction.

### Questions about this correction for @neukym 

Is this correct? Is this correcting probability, or one-minus-probability as suggested in the variable name?  It looks to me like `probability_X_gt_x` should be the upper tail of the normal distribution, which is to say it's (depending on how you look at it) the p threshold not 1-p threshold.  So I think the variable name is a little misleading, and that the signature is otherwise correct.

As another point of comparison, on main currently there's also a snippet in `plot/plot.py` doing sidak correction for the alpha threshold:

```python
    sidak_corrected_alpha = 1 - (
        (1 - alpha)
        ** np.longdouble(1 / (2 * len(expression_set.latencies) * n_channels * len(show_only)))
    )
```
(this code recently touched by @young-x-skyee but just to change the float to a longdouble).  So here it looks like the correction is applied to the alpha level, not 1-p.  On branch, this is refactored into a single function in one place, assuming that it's working on p.

### How much correction to apply

Of particular note however is that the number of timepoints and hexels are stored in constants in `constants.py`:

```python
TIMEPOINTS = 201
NUMBER_OF_HEXELS = 200000
```

(These were previously stored in vars in a few places, so I just put them in one place.)

My understanding is that the denominator in the exponent, everything after the `2 *` (i.e. `TIMEPOINTS * NUMBER_OF_HEXELS` in the IPPM code and `len(expression_set.latencies) * n_channels * len(show_only)` in the expression plot code) is the number of comparisons we're correcting for.

It would be nice if these values were derived from the data rather than constants, and I've tried to do this on branch. However if this is true than I think the constants are wrong...
```python
len(expression_set.latencies)  # 201 - that's correct
len(expression_set.hexels_left) + len(expression_set.hexels_right)  # 20484
```
So 20,484 is out from `NUMBER_OF_HEXELS = 200_000` by a factor of 10.

Another point is also raised comparing to the Šidák correction in the expression plotting code: this includes the number of transforms in the correction factor (that's `len(show_only)` in that code).  In trying to derive the IPPM threshold from the data, I have included that logic.  Here's the IPPM threshold on branch:

```python
            def get_threshold(es: ExpressionSet) -> float:
                """Get Šidák n-sigma threshold"""
                n_comparisons = len(es.transforms) * len(es.latencies) + get_n_channels(es)
                return p_to_logp(sidak_correct(p_threshold_for_sigmas(exclude_points_above_n_sigma),
                                               n_comparisons=n_comparisons))
```
```python
def sidak_correct(alpha: float, n_comparisons: int) -> float:
    """Applies Šidák correction to an alpha threshold."""
    return 1 - (
        (1 - alpha)
        ** longdouble(1 / (2 * n_comparisons))
    )
```
(just converting to logp values here because that's how probability is stored in expression sets).

But then my question is: am I right to do this? On main, the IPPM code doesn't adjust the threshold depending on how many transforms are in play - but should it?  Or should we treat each transform entirely independently?

On main, about 86% of the points are removed using this threshold before denoising even begins, so it makes a big difference to what's going on.

## Conclusion

The branch code is obviously doing something wrong, but given that by some argument a threshold *should* change depending on the number of transforms included in the IPPM, and that interacts with the behaviour observed on branch, I think we should resolve this before continuing.

### TL;DR

1. Is the Šidák correction on branch out by an order of magnitude, and if so, would this change the recommendations of the paper?
2. Should the correction take into account the number of transforms being included in the IPPM?

**Comment by caiw** [comment]
(And of course I hasten to add that if there is an error on main, it's entirely something which @neukym and I should have picked up in review!)

**Comment by neukym** [comment]
Yes I agree.

> Bonus question for @neukym: what's that 2 doing there? If that's 2 hemispheres I don't think we need it if we just count all the hexels. Thoughts?

Yes, you are right, the 2 is there for the number of hemispheres.

> Is this correct? Is this correcting probability, or one-minus-probability as suggested in the variable name?

Good catch. The correct one is on the plot.py:

```
  sidak_corrected_alpha = 1 - (
        (1 - alpha)
        ** np.longdouble(1 / (2 * len(expression_set.latencies) * n_channels * len(show_only)))
    )
 ```
 
...although as you point out the `2` is **erroneous** here if `n_channels` can be either the total hexels or sensors - it is a legacy from when this was 'number of hexels per hemisphere'. Your `sidak_correct` in probability.py is a good solution I think - both plot.py, ippm.py etc should refer back to this one function:

```
def sidak_correct(alpha: float, n_comparisons: int) -> float:
    """Applies Šidák correction to an alpha threshold."""
    return 1 - (
        (1 - alpha)
        ** longdouble(1 / (n_comparisons))
    )
```

where `n_comparisons = n_timepoints * n_transforms * n_channels`. **Note I have just removed the erroneous `2` from this function myself in commit 300662e.**

> @neukym Šidák correction on main currently doesn't take into account the number of transforms being included in the IPPM — but should it?

Yes I think it should - we take account of this in main's `plot.py`, and this is the one we should be using.

> @neukym @anirudh1666 Is the Šidák correction on branch out by an order of magnitude, and if so, would this change the recommendations of the paper?

Yes, you are right that the following logic is correct:

```
len(expression_set.latencies)  # 201 - that's correct
len(expression_set.hexels_left) + len(expression_set.hexels_right)  # 20484
```

`NUMBER_OF_HEXELS = 200000` is incorrect, and you are right to discard it. 👍

> @neukym Šidák correction on main currently doesn't take into account the number of transforms being included in the IPPM — but should it?

Yes, I think it should. To standardise things, perhaps we should add this function to `probability.py`, and then always use this?:

```
def calculate_alpha_threshold(n_timepoints, n_channels, n_transforms):
   n_comparisons = n_timepoints * n_channels* n_transforms
   alpha = p_threshold_for_sigmas(5) # use 5-sigma
   alpha_threshold =  sidak_correct(alpha, n_comparisons)
   return alpha_threshold
   
Where

```
n_timepoints = len(expression_set.latencies)  # 201 - that's correct
n_channels = len(expression_set.hexels_left) + len(expression_set.hexels_right)  # 20484
n_transforms = len(show_only)
```

This then standardises everything into a single function, and it forces everyone to explicitly put in the number of time points, channels and transforms into it. (I'm worried if we don't make this function explicit there's a risk that one of us will forget to put in the number of transforms, as has been happening so far.)

**The problem**

I think I am right in saying the denoiser uses this threshold to do an initial filter, before it starts clustering. If we were to use the above logic this would mean that the denoised expression dataset is specific to the scope of some specific set of transforms in show_only, which doesn't seem quite right. We perhaps need to chat about this - I hadn't thought of this situation. 

One `hack` would be not to include n_transforms, in the denoiser, so for that, in `DenoisingStrategy.py` the input values for `calculate_alpha_threshold` would be:

```
n_timepoints = len(expression_set.latencies)  # 201 - that's correct
n_channels = len(expression_set.hexels_left) + len(expression_set.hexels_right)  # 20484
n_transforms = 1
```

We then reapply: 

```
n_timepoints = len(expression_set.latencies)  # 201 - that's correct
n_channels = len(expression_set.hexels_left) + len(expression_set.hexels_right)  # 20484
n_transforms = len(show_only)
```
in the IPPM plotting step. Bit hacky though.

**Comment by caiw** [comment]
Hey, I've not been able to get to this properly today for half-term childcare reasons, but I have some thoughts. I'll get back to it properly next week.

## Alpha-level correction

@neukym said:
> **Note I have just removed the erroneous `2` from this function myself in commit [300662e](https://github.com/kymata-atlas/kymata-core/commit/300662e2372016901ac9c7a757877571a5f42339).**

Great, I'll use this, and I've added some tests for specific values, using an online Sidak calculator to verify the correct results.

@caiw said:
> Šidák correction on main currently doesn't take into account the number of transforms being included in the IPPM — but should it?

@neukym said:
> Yes, I think it should. To standardise things, perhaps we should add this function to `probability.py`, and then always use this?

We could, but I think we should push this to another PR. In particular we describe Sidak correction in the plot axis name for expression plots, so I'm slightly unsure about also setting that choice elsewhere. In particular, if some reviewer some time demands bonferroni correction, or we decide that one correction or another makes sense at some point, I wonder if we should keep it a bit more procedural.

I don't feel strongly on this point.

@anirudh1666 said:
> But when we do denoising, the threshold for significance changes based on the subset size. This intuitively feels wrong because selecting a subset of the results should not change the results?

I know what you mean, but it's all about controlling the chance that we find false signal in the noise. The more things we test, the more likely things are likely to appear significant by chance.  The alpha level is precisely the level of "false discoveries" we are willing to accept.  So the more transforms we let in into a given "omnibus test" (testing multiple things all at the same time), the more likely one of them is likely to pass a threshold by chance, so the more we have to restrict ("correct") the threshold for significance to control the overall false-discovery rate) — I think each transform we add (and every latency and hexel) is like an [extra jellybean colour](https://xkcd.com/882/), which is why we have to control the alpha level so aggressively.

@neukym is the expert on the stats comparisons made in the kymata gridsearch, and under certain circumstances we can control the FDR without correcting a threshold in this brute-force way, but doing the correction is the safest/most conservative way.  Unless @neukym would argue the model comparison we do is somehow automatically controlling the familywise error rate (I've not thought about this in full), I think we should do full correction for all comparisons.

@neukym said:
> I think I am right in saying the denoiser uses this threshold to do an initial filter, before it starts clustering. If we were to use the above logic this would mean that the denoised expression dataset is specific to the scope of some specific set of transforms in show_only, which doesn't seem quite right. We perhaps need to chat about this - I hadn't thought of this situation - that the denoising procedure would be reliant on the value in `show_only`.

So I've changed my mind on this. We're using an alpha threshold to apply an initial filter of points passed to the denoiser, but we're not actually using it for a formal omnibus NHST, so in that sense we don't *need* to control the FWER in a strict sense like we do when presenting an expression plot (where the coloured points are RESULTS so we *need* to control the FWER in those results).  So long as the significance of the nodes in the final IPPM survive a fully corrected NHST, I think we are fine applying an independent threshold to each transform going into the denoiser.

To be clear, my recommendation is the same as @neukym's:

1. Apply a fixed (`n_transforms = 1`) threshold to the `ExpressionSet` — this is essentially to help the denoiser, so it doesn't matter if it's not fully corrected.
2. Apply the denoiser.
3. Apply a corrected (`n_transforms = len(denoised_clusters)`) threshold to the surviving clusters to ensure we don't have uncontrolled FWER.

If this happens within the IPPM-building step, it can be documented in code, and the thresholds should\* be quick to calculate and apply.

(\*As @anirudh1666 points out there is currently a bug in this, but that's easily fixable)

@neuky said:
> The other, cleaner, solution is that denoising simply isn't callable _unless you are plotting an IPPM_ (in which case `show_only`will be available). This would require the denoiser to work quite quickly, so it can be done on the fly each time you plot an IPPM.

I don't favour this exactly — there are other reasons we may want to denoise multiple transforms in future which don't directly result in IPPMs (e.g. see what Tianyi is doing with thousands of related transforms from ANN nodes). As previously discussed, I think we should formalise denoising as a process *on `ExpressionSet`s*, and decouple it from the process of drawing an IPPM from a denoised `ExpressionSet`.

@anirudh1666 said:
> My understanding is that with the new threshold, the denominator will increase, so sidak_corrected_alpha will be lower. Which in turn should raise the threshold for significance? (May be wrong)

It's the other way: on `main` we're correcting for ~10x the comparisons which means the threshold will be lower, because α << 1. (Lower in absolute terms — always confusing with p-values, -log p-values, etc).  So on `main` the number of points going into the denoiser is much smaller than it will be having fixed the correction bug on branch.

@anirudh1666 said:
> Consequently, some transforms which had significant nodes may no longer have any significant nodes. It wouldn’t change the location of nodes or the most significant node

Yeah, that makes sense to me — good point.

## Refactoring `ExpressionSet`

@anirudh1666 said:
> - **Refactor ExpressionSet, maybe some of DenoisingStrategy.** The expression set class is quite big, so I think it would be good to refactor it into small modules that abstract away some of the complexity, which makes it more readable and debuggable. Also, will add some pretty printing functionality as it can be hard to trace the state of ExpressionSets using solely print statements. This will take time though as I need to build an in-depth understanding of ExpressionSet and find good cut points.

I'm loath to do this lightly. We spent quite some time building this class to be performant in terms of space and time, and also future-proof in terms of possible additional data sources, e.g. OPM, EEG, ECoG.

In particular, compared to what the kymata.org API serves, which is what the IPPM functions were originally coded against, `ExpressionSet`s contain *a lot more information* (i.e. the p-values of all the non-best transforms at each datapoint), but we do use that additional information in our analyses.

In particular `ExpressionSet`s do model selection — i.e. always showing the best model at each datapoint from the available set — automatically, as part of the data model.  In the API version, this model selection is done *prior* to the data model, and as such ad-hoc model selection is impossible — a batch of results from the API is what it is and can't be explored in the same way an `ExpressionSet` can.

Similarly...

@anifudh1666 said:
>  I would have also preferred starting with a simpler version of ExpressionSet that is correct and works, then iteratively optimise and generalise it, making sure the tests/demo_ippm are working after each iteration.

Yeah, I fully mea-culpa that this is a monster refactor and that's not nice for anyone and has made the testing hard — sorry!

I will say though that `ExpressionSet` has been around for ages as a class and is used for a lot of disk i/o as well as the main kymata gridsearch, so one goal of this PR was rather to align the IPPM code with the core data model of the rest of `kymata-core`, as opposed to maintaining a separate, parallel data model for purposes of denoising.

We previously (in the Matlab version of the code around for years) had a lot more of a simple/modular/single-user version of `ExressionSet` called "endtables", which caused us a lot of problems.  In particular, the ability to split and combine `ExpressionSet`s computed in different experimental runs, and have them validate the ligitimacy of this operation, and perform flexible model selection as part of this split/combine operations is really important, and in our previous more modular approach this required a *lot* of user bookkeeping and was a source of errors and mental effort.  It was dependent on the user to keep track of what functions had been added and removed from the endtable, especially when saving and loading data to disk, rather than the class itself automatically tracking this.

IMO, the mathematical framework we're trying to describe in the class here is just inherently rather complex, and that complexity is admittedly reflected in the class.  However previously when we had a simpler class, it actually just pushed that complexity onto the user, requiring them to do things in a certain order and keep track of things using things like choice of file names (😱) and the top-level script, which meant an actually fairly deep understanding of the underlying maths was required to even write a top-level script which behaved correctly.  What we pay for in class complexity here we to some degree get back in "automatic correctness" of the behaviour.

I guess what I'm saying is that I'm not sure the optimisation is as premature as it may appear :)

### But

**I'm absolutely *not* saying that these classes can't be improved!**  What I'm saying is I would recommend we consider this as a separate issue, and not a requirement of aligning the IPPM code with the `ExpressionSet` class as-is (i.e. this PR).

### Having said that...

@anirudh1666 said:
> There may be some problems in ExpressionSet._clear_points

and

> There are other problems too, for example, shuffling the points changes the coordinates in the sparse array

This `.clear_points` functionality is new as of this PR, so it is not as well battle-tested as the rest of the class, and I am sure you're right as to these bugs. This is the priority for me to fix, and I think is *the* problem.

## Next

Rather than @anirudh1666 spending ages wrestling with the unfamiliar `ExpressionSet` class, I think I should take another shot at stepping the tests through in `main` vs `branch` to see where the broken integration tests are actually failing.  I believe. Anirudh is right that it's in `.clear_points` and shuffling, and clearly the behaviour is wrong, so that's where to look.  @neukym happy to chat about this in term of general priorities but maybe that's the next step for this PR?

**Comment by caiw** [comment]
Anyway, long comment finally posted now I have internet access again. Thanks @neukym and @anirudh1666 for the thoughtful comments!

**Comment by neukym** [comment]
Thanks both - great discussion.

**Comment by anirudh1666** [comment]
> I know what you mean, but it's all about controlling the chance that we find false signal in the noise. The more things we test, the more likely things are likely to appear significant by chance. The alpha level is precisely the level of "false discoveries" we are willing to accept. So the more transforms we let in into a given "omnibus test" (testing multiple things all at the same time), the more likely one of them is likely to pass a threshold by chance, so the more we have to restrict ("correct") the threshold for significance to control the overall false-discovery rate) — I think each transform we add (and every latency and hexel) is like an [extra jellybean colour](https://xkcd.com/882/), which is why we have to control the alpha level so aggressively.

This makes sense to me. Unfortunately, the stuff that came next went over my head :D. This might be annoying but I still want to understand what the problem is here. The threshold for significance is based on the sample size because the more results we generate, the greater the probability of a false positive (as Cai has explained in the quote above). The problem is that the threshold we are estimating in the denoiser is based on a subset of the transformations, so it does not accurately reflect the corrected threshold for the entire dataset. This mismatch in thresholds is causing the problem? By doing,

> 1. Apply a fixed (n_transforms = 1) threshold to the ExpressionSet — this is essentially to help the denoiser, so it doesn't matter if it's not fully corrected.
> 2. Apply the denoiser.

we use the threshold that is estimated from the entire dataset, i.e., we don't estimate a new threshold in the denoiser but use the n_transforms = 1 threshold. But the last step, 

> 3. Apply a corrected (n_transforms = len(transforms)) threshold to the surviving clusters to ensure we don't have uncontrolled FWER.

doesn't make sense to me. Surely the threshold with n_transforms = 1 is more strict because it has more results, so applying this threshold shouldn't make a difference? I may have misunderstood or made a false assumption somewhere.

> So on main the number of points going into the denoiser is much smaller than it will be having fixed the correction bug on branch.

Yep, this makes sense. With this explanation, I think this will lead to more nodes appearing because there may be more clusters that were previously discarded because they were below the significance threshold. However, the hyperparameter grid search suggested global max pooling is optimal. If this strategy is actually optimal, then the results won't change because the maximum wont change.

**Comment on refactoring expression set**

I definitely agree with the current version of ExpressionSet being very powerful and flexible, the complexity being essential rather than superfluous, and the issues with exposing low-level details to the user. My suggestion was more along the lines of organising some of this functionality to improve readability rather than removing it. I.e., the behaviour of ExpressionSet remains exactly the same, it is just organised differently behind the scenes. Some of the complexity could be pushed into lower level modules, which would lead to faster debugging time. Here is one example: in the constructor, we perform a lot of useful guardrail checks which should be kept. However, in this context, I could safely assume that they weren't causing any denoising problems but since this complexity is naked in the constructor, I had to go through it incase I missed the odd line that was unrelated to guardrail checks. We could abstract these guardrails behind a well-named function, so the programmer can read the name, guess what it does, and decide whether they need to read the code or they can skip past it. However, I agree this should be done in a different pull request. The first step should be to get the current version working. Refactoring ExpressionSet would be a good programming exercise, so I am interested in coming up with a list of suggestions and we can decide which are worth pursuing.

The reason I am insisting on readability is that (I can't find the original paper where I read this unfortunately but they essentially tracked the mouse movements of programmers for an hour) the ratio of time spent reading code to writing is 10:1. In other words, writing code that is easy to read massively speeds up developer productivity. 

**Next Steps**

I am happy with your suggestion. I want to call out that there are other issues besides shuffling as well, most of the preprocessing modifies the data, which leads to inconsistent state between ExpressionSet and the DenoisingStrategy data. I tried turning off all of the preprocessing/extra steps but it didn't fix the issue, so there must be something else. I will commit a working DenoisingStrategy integration test I have. The other integration tests will have the same pattern, which different inputs and hyperparameters. This should also let you test your changes quickly because you can run this test in isolation rather than having to run demo_ippm.ipynb every time. 

Link to working integ test: https://github.com/kymata-atlas/kymata-core/blob/ippm-graph-refactor-integ-test/tests/test_ippm_denoising_strategies_unit_test_new.py

**Comment by caiw** [comment]
I'll write some comments as I go and then write "DONE" when I'm done writing comments.

**Comment by caiw** [comment]
@anirudh1666 said:

> There may be some problems in ExpressionSet._clear_points

Yep, I've confirmed that the denoising is actually working, and it's `ExpressionSet._clear_points` which is not working.

**Comment by caiw** [comment]
@anirudh1666 said:
> the last step,
>> Apply a corrected (n_transforms = len(transforms)) threshold to the surviving clusters to ensure we don't have uncontrolled FWER.
>
> doesn't make sense to me. Surely the threshold with n_transforms = 1 is more strict because it has more results, so applying this threshold shouldn't make a difference? I may have misunderstood or made a false assumption somewhere

The way I'm thinking about it, assume we have just one single transform. In a standard non-denoised expression plot, we do the following individual statistical tests:
- We perform a test for every hexel in the brain
- For each of those we test every latency in our chosen range

So the total number of tests we do is $n_{H} \cdot n_{L}$, and so if we're setting our p-threshold at some standard for significance testing like $p<\alpha = 0.05$, each one of of those $n_{H} \cdot n_{L}$ tests has a probability of $\alpha$ of passing the test even if there's no true effect. (There's lots of assumptions here about independence of tests which aren't satisfied, so again this is just being ultra conservative.) The most brute-force way of correcting this is to divide $\alpha$ by the number of tests we're doing, so instead our new test would be to set a corrected threshold $p < \alpha* = 0.05 / n_{H} \cdot n_{L}$ — that's Bonferroni correction, and it ensures that the *overall* chance of *any* of those tests being a false positive is controlled at 0.05.

If we add a second transform to the mix, we're repeating each of those tests for the second transform: at a given latency and a given hexel, even if transform 1 is insignificant, we also check transform 2.  So now the total number of individual tests is $n_{H} \cdot n_{L} \cdot 2$, and so on up to $n_{H} \cdot n_{L} \cdot n_{T}$.

Again it's actually more complex: we do model selection, so worse-fitting transforms are excluded before they're tested, but by adding a new transform into the mix we're still increasing the chance that we get a false positive because we're performing *some* additional tests, so just correcting for the total number is the most conservative method.

The only difference in what we're doing is that our correction is Sidak correction instead of Bonferroni correction, so instead of

$\alpha* = \frac{\alpha}{n_{H} \cdot n_{L} \cdot n_{T}}$

we do

$\alpha* = \alpha^{\frac{1}{n_{H} \cdot n_{L} \cdot n_{T}}}$

But the logic is the same: number of tests goes up, $\alpha*$ goes down to ensure we don't raise the likelihood that we get a false positive by just doing more tests until one comes out positive.

When the denoiser is being applied to one transform, it makes sense to treat that one alone, however when the results of all denoised transforms are combined, we're effectively adding an extra batch of tests for each transform, so the final clusters should be tested against $\alpha*$ as above.  Again treating each transform as independent is over-conservative, but I don't think we can not apply correction for the extra transforms.

**Comment by caiw** [comment]
@anirudh1666 said:

> My suggestion was more along the lines of organising some of this functionality to improve readability rather than removing it. I.e., the behaviour of ExpressionSet remains exactly the same, it is just organised differently behind the scenes. 
> [...]
> Here is one example: in the constructor, we perform a lot of useful guardrail checks which should be kept. However, in this context, I could safely assume that they weren't causing any denoising problems but since this complexity is evident in the constructor, I had to go through it incase I missed the odd line that was unrelated to guardrail checks. We could abstract these guardrails behind a well-named function, so the programmer can read the name, guess what it does, and decide whether they need to read the code or they can skip past it.

Ok, yes, this is a really good example and I totally agree with this :)

> Refactoring ExpressionSet would be a good programming exercise, so I am interested in coming up with a list of suggestions and we can decide which are worth pursuing.

Please do!  You're totally right about ease of readability being of paramount importance, especially in a lab which will hopefully be having new students coming across the code afresh from time to time.  Any suggestions you have about ways to increase the readability would be very welcome :)

**Comment by caiw** [comment]
So after a chat with @neukym this afternoon I think we're 99% of the way there. Some of the tests not passing is due to the threshold being different, so my first goal is basically to get the demo to work through.  I've fixed the bug with the non-denoised points not being cleared properly, and now I just need to iron out some visual stuff.

**Comment by caiw** [comment]
@caiw said:
> I'll write some comments as I go and then write "DONE" when I'm done writing comments.
DONE

**Comment by neukym** [comment]
> Could we reduce `n_H⋅n_L` down to just `n_clusters`? What do you think @neukym?

Interesting - yes maybe. But can we leave this for the moment - let's standardise everything around `n_H⋅n_L` first.

**Comment by anirudh1666** [comment]
Thanks for your in-depth explanation regarding the test for significance. I am able to follow your reasoning. 

So, the problem here is that we pass in a subset of transforms into the denoiser to estimate the threshold for significance, and n_t (where n_t = number of transforms) is equal to the subset cardinality rather than the entire set. Therefore, the threshold is less conservative and has a greater chance of allowing false positives. 

If my understanding is correct, could we solve this by setting the threshold for significance in the denoiser to be the full set cardinality rather than the subset? Then, the threshold for significance remains the same and is the most conservative value. Why would we modify this threshold AFTER the denoising? Is it to help the denoiser performance? I could plausibly see that having a large number of insignificance points around a significant spike could help the denoiser tag it as significant.

We could make the threshold for significance a hyperparameter, and so, configurable by the user. Then, the denoiser will be unaware of this knowledge, and it should be easier to change.

**Comment by caiw** [comment]
@neukym requested to merge this without another round of review. See #441 for remaining tests to fix
